File: swift/number2.swift
// Swift
import Foundation

let forbiddenCharsSwift = CharacterSet(charactersIn: "&=+<>,_'-")
let allowedCharsBaseSwift = CharacterSet.alphanumerics.union(CharacterSet(charactersIn: ".*"))

// Проверка имени пользователя на соответствие правилам
func isValidUsernameSwift(username: String) -> Bool {
    guard username.count >= 6 && username.count <= 30 else { return false }
    guard username.rangeOfCharacter(from: forbiddenCharsSwift) == nil else { return false }
    guard !username.hasPrefix(".") && !username.hasSuffix(".") else { return false }
    guard !username.contains("..") else { return false }
    guard username.rangeOfCharacter(from: allowedCharsBaseSwift.inverted) == nil else { return false }
    return true
}

func emailCountSwift(emailList: [String]) -> Int {
    // Используем Set для хранения уникальных нормализованных адресов
    var uniqueEmails = Set<String>()

    for email in emailList {
        guard let atIndex = email.firstIndex(of: "@"),
              atIndex > email.startIndex,
              atIndex < email.index(before: email.endIndex)
        else {
            continue
        }

        var local = String(email[..<atIndex])
        let domain = String(email[email.index(after: atIndex)...])

        // Сначала валидируем оригинальное имя пользователя
        guard isValidUsernameSwift(username: local) else {
            continue
        }

        // Нормализация: убираем все после '*' и удаляем все '.'
        if let starIndex = local.firstIndex(of: "*") {
            local = String(local[..<starIndex])
        }
        let cleanedLocal = local.replacingOccurrences(of: ".", with: "")

        uniqueEmails.insert(cleanedLocal + "@" + domain)
    }
    return uniqueEmails.count
}

// Example Usage:
let addressList1 = ["mar.pha@corp.nstu.ru", "marpha@corp.nstu.ru", "marph.a@corp.nstu.ru"]
let addressList2 = ["mar.pha*science@corp.nstu.ru", "marpha*scie.nce@corp.nstu.ru", "marph.a*s.c.i.e.n.c.e@corp.nstu.ru"]
let addressList3 = ["mar.pha*science@co.rp.nstu.ru", "marpha*scie.nce@corp.nstu.ru", "marph.a*s.c.i.e.n.c.e@corp.nstu.ru"]

print("List 1 Count: \(emailCountSwift(emailList: addressList1))")
print("List 2 Count: \(emailCountSwift(emailList: addressList2))")
print("List 3 Count: \(emailCountSwift(emailList: addressList3))")

File: go/number2.go
package main

import (
	"fmt"
	"strings"
)

// Функция для проверки корректности имени пользователя
func isValidUsername(username string) bool {
	if len(username) < 6 || len(username) > 30 {
		return false
	}
	if strings.HasPrefix(username, ".") || strings.HasSuffix(username, ".") {
		return false
	}
	if strings.Contains(username, "..") {
		return false
	}
	return true
}

func emailCount(emailList []string) int {
	uniqueEmails := make(map[string]struct{})

	for _, email := range emailList {
		parts := strings.Split(email, "@")
		if len(parts) != 2 {
			fmt.Println("Некорректный email:", email)
			continue
		}

		local := parts[0]
		domain := parts[1]

		if !isValidUsername(local) {
			fmt.Println("Некорректное имя пользователя:", local)
			continue
		}

		local = strings.ReplaceAll(local, ".", "")

		if starIndex := strings.Index(local, "*"); starIndex != -1 {
			local = local[:starIndex]
		}

		uniqueEmails[local+"@"+domain] = struct{}{}
	}

	return len(uniqueEmails)
}

func main() {
	// Пример 1
	addressList1 := []string{
		"mar.pha+science@corp.nstu.ru",
		"marpha+scie.nce@corp.nstu.ru",
		"marph.a+s.c.i.e.n.c.e+@corp.nstu.ru",
	}
	fmt.Printf("Output (Пример 1): %d\n", emailCount(addressList1))

	// Пример 2
	addressList2 := []string{
		"mar.pha+science@co.rp.nstu.ru",
		"marpha+scie.nce@corp.nstu.ru",
		"marph.a+s.c.i.e.n.c.e+@corp.nstu.ru",
	}
	fmt.Printf("Output (Пример 2): %d\n", emailCount(addressList2))

}